backend/LeetWars.Core/LeetWars.Core.BLL/Services/ChallengeService.cs

```diff
@@ -40,27 +56,85 @@ public async Task<ICollection<ChallengePreviewDto>> GetChallengesAsync(Challenge
                     challenge.Versions.Any(version => version.LanguageId == filters.LanguageId));
             }
 
-            if (filters.TagsIds != null)
-            {
-                challenges = challenges.Where(challenge =>
-                    challenge.Tags.Any(tag => filters.TagsIds.Contains(tag.Id)));
-            }
-
             if (filters.Progress.HasValue)
             {
                 challenges = filters.Progress switch
                 {
-                    ChallengesProgress.NotStarted => challenges.Where(challenge => !challenge.Versions.Any(version => version.Solutions.Any())),
-                    ChallengesProgress.Started => challenges.Where(challenge => challenge.Versions.Any(version => version.Solutions.Any())),
-                    ChallengesProgress.Completed => challenges.Where(challenge => challenge.Versions.All(version => 
-                        version.Solutions.All(solution => solution.SubmittedAt.HasValue && solution.SubmittedAt.Value != DateTime.MinValue))),
+                    ChallengesProgress.NotStarted => challenges.Where(challenge => challenge.Versions.All(version =>
+                        version.Solutions.Count == 0 || version.Solutions.All(solution =>
+                            solution.User == null || solution.User.Uid != userId))),
+                    ChallengesProgress.Started => challenges.Where(challenge => challenge.Versions.Any(version =>
+                        version.Solutions.Any(solution =>
+                            solution.User != null && solution.User.Uid == userId && solution.SubmittedAt == null))),
+                    ChallengesProgress.Completed => challenges.Where(challenge => challenge.Versions.Any(version =>
+                        version.Solutions.Any(solution =>
+                            solution.User != null && solution.User.Uid == userId && solution.SubmittedAt.HasValue &&
+                            solution.SubmittedAt.Value != DateTime.MinValue))),
```

Comment:
Submitted date can be null or value?
how can be minvalue here?

Reply:
overinsurance, removed